src/scripts/badgeCase/BadgeCase.ts

```diff
@@ -50,15 +39,14 @@ class BadgeCase implements Feature {
             return;
         }
 
-        for (let i = 0; i < this.badgeList.length; i++) {
-            this.badgeList[i] = json[i];
-        }
+        json.forEach((hasBadge, index) => {
+            this.badgeList[index] = hasBadge;
+        });
     }
 
     toJSON(): Record<string, any> {
-        return this.badgeList.map(badge => {
-            return badge;
-        });
+        let shouldReturn = false;
+        return [...this.badgeList].reverse().filter(i => shouldReturn || i && (shouldReturn = i)).reverse();
```

Comment:
Might just be me, but this isn't entirely clear. Seems to be so we only store what is necessary, ignoring after the last acquired badge.
I can't think of a solution which is immediately clearer, so maybe we should just add a comment here saying what it is supposed to be doing.